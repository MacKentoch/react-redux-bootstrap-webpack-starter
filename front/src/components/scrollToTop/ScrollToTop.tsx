import React, { Fragment, ReactNode } from 'react';
import { useLocation } from 'react-router-dom';
import { useScrollToTopOnLocationChange } from './hooks/useScrollToTopOnLocationChange';
type OwnProps = {
  children: ReactNode;
};
type Props = OwnProps;

function ScrollToTop({ children }: Props) {
  const location = useLocation();
  useScrollToTopOnLocationChange(location);

  return <Fragment>{children}</Fragment>;
}

ScrollToTop.displayName = 'ScrollToTop';

export default ScrollToTop;
