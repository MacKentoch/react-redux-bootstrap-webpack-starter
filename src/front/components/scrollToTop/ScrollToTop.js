// @flow

import React, { Fragment, useEffect, useRef } from 'react';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import { withRouter } from 'react-router';

// #region flow types
type Props = {
  // react-router 4:
  match: Match,
  location: Location,
  history: RouterHistory,

  children: any,
};
// #endregion

function useScrollToTopOnLocationChange(location: any) {
  const prevLocation = useRef();

  useEffect(() => {
    prevLocation.current = location;
  }, []);

  useEffect(() => {
    if (prevLocation.current !== location) {
      window && window.scrollTo(0, 0);
      prevLocation.current = location;
    }
  }, [location]);
}

function ScrollToTop({ children, location }: Props) {
  useScrollToTopOnLocationChange(location);

  return <Fragment>{children}</Fragment>;
}

ScrollToTop.displayName = 'ScrollToTop';

// $FlowIgnore
export default withRouter(ScrollToTop);
