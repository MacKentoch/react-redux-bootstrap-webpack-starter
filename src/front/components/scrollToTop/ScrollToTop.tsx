import React, { Fragment, useEffect, useRef } from 'react';
import { withRouter, RouteComponentProps } from 'react-router';

// #region types
type Props = {
  children: any,
} & RouteComponentProps;
// #endregion

function useScrollToTopOnLocationChange(location: any) {
  const prevLocation = useRef();

  useEffect(() => {
    prevLocation.current = location;
  }, []);

  useEffect(() => {
    if (prevLocation.current !== location) {
      window && window.scrollTo(0, 0);
      prevLocation.current = location;
    }
  }, [location]);
}

function ScrollToTop({ children, location }: Props) {
  useScrollToTopOnLocationChange(location);

  return <Fragment>{children}</Fragment>;
}

ScrollToTop.displayName = 'ScrollToTop';

export default withRouter(ScrollToTop);
