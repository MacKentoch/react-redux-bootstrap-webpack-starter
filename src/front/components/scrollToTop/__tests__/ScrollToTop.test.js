// @flow

// #region imports
import React from 'react';
import renderer from 'react-test-renderer'; // needed both for snpashot testing but also to prevent errors from enzyme
import { MemoryRouter } from 'react-router';
import ScrollToTop from '../ScrollToTop';
// #endregion

describe('ScrollToTop component', () => {
  it('renders as expected', () => {
    const component = renderer
      .create(
        <div>
          <MemoryRouter>
            <ScrollToTop>
              <p>a child</p>
            </ScrollToTop>
          </MemoryRouter>
        </div>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
