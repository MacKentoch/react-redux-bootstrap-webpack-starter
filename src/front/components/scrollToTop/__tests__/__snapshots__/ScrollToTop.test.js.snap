// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTop component renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div>
    <MemoryRouter>
      <withRouter(ScrollToTop)>
        <p>
          a child
        </p>
      </withRouter(ScrollToTop)>
    </MemoryRouter>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <MemoryRouter>
        <withRouter(ScrollToTop)>
          <p>
            a child
          </p>
        </withRouter(ScrollToTop)>
      </MemoryRouter>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <withRouter(ScrollToTop)>
          <p>
            a child
          </p>
        </withRouter(ScrollToTop)>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <p>
            a child
          </p>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "a child",
          },
          "ref": null,
          "rendered": "a child",
          "type": "p",
        },
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <MemoryRouter>
          <withRouter(ScrollToTop)>
            <p>
              a child
            </p>
          </withRouter(ScrollToTop)>
        </MemoryRouter>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <withRouter(ScrollToTop)>
            <p>
              a child
            </p>
          </withRouter(ScrollToTop)>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <p>
              a child
            </p>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "a child",
            },
            "ref": null,
            "rendered": "a child",
            "type": "p",
          },
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
